File: 1679-max-number-of-k-sum-pairs/1679-max-number-of-k-sum-pairs.py
class Solution:
    def maxOperations(self, nums: List[int], k: int) -> int:
        """
        1 <= nums[i] <= 10^9
        """
        arr = [i for i in nums if i < k]
        print(arr)
        
        arr = sorted(arr)
        
        sol = 0
        i, j = 0, len(arr)-1
        while i<j:
            if arr[i] + arr[j] > k:
                j-=1
            elif arr[i] + arr[j] < k:
                i+=1
            else:
                sol += 1
                i, j = i+1, j-1
                
        return sol
            

File: 1679-max-number-of-k-sum-pairs/1679-max-number-of-k-sum-pairs.cpp
class Solution {
public:
    bool predicate(int element, int k) {
        return (element < k);
    }
    int maxOperations(vector<int>& nums, int k) {
        vector<int> fnums(nums.size());
        copy_if(nums.begin(), nums.end(), fnums.begin(), [=](int em) {
            return predicate(em, k);
        });
        
        // sort
        sort(fnums.begin(), fnums.end());
        /* debug
        for_each(fnums.begin(), fnums.end(),
                  [](const int& n) { cout << n << " "; });
        cout << endl;
        */

        int sol = 0;
        int i = 0;
        int j = fnums.size() - 1;
        while (i<j) {
            if (fnums[j] > k or fnums[i] + fnums[j] > k) {
                j--;
            } else if (fnums[i] + fnums[j] < k) {
                i++;
            } else {
                sol++;
                i++;
                j--;
            }
        }
        return sol;
    }
};